src/lib/datepicker/datepicker-input.ts

```diff
@@ -123,38 +123,37 @@ export class MatDatepickerInput<D> implements ControlValueAccessor, OnDestroy, V
 
   /** The value of the input. */
   @Input()
-  get value(): D | null { return this._value; }
+  get value(): D | null { return this._value? this._value.asDate(): null; }
   set value(value: D | null) {
-    value = this._dateAdapter.deserialize(value);
-    this._lastValueValid = !value || this._dateAdapter.isValid(value);
-    value = this._getValidDateOrNull(value);
-    const oldDate = this.value;
-    this._value = value;
-    this._formatValue(value);
-
-    if (!this._dateAdapter.sameDate(oldDate, value)) {
-      this._valueChange.emit(value);
+    const oldDate = this._value;
+    this._value = new MatSingleDateSelection(this._dateAdapter, value);
+    this._lastValueValid = this._value.isValid(); // TODO(sgoldblatt): is this right?
+    
+    this._formatValue(this._value.asDate());
+
+    if (!this._value.isSame(oldDate)) {
+      this._valueChange.emit(this._value.asDate());
     }
   }
-  private _value: D | null;
+  private _value: MatSingleDateSelection<D>;
 
   /** The minimum valid date. */
   @Input()
-  get min(): D | null { return this._min; }
+  get min(): D | null { return this._min? this._min.asDate(): null; }
   set min(value: D | null) {
-    this._min = this._getValidDateOrNull(this._dateAdapter.deserialize(value));
+    this._min = new MatSingleDateSelection(this._dateAdapter, value);
     this._validatorOnChange();
   }
-  private _min: D | null;
+  private _min: MatSingleDateSelection<D>;
```

Comment:
The `min` and `max` can stay just dates, even in a range selector we would expect the min and max to be dates rather than ranges

Reply:
mk

Reply:
hmm actually I looked back and this and we compare using the min and max so I think the matsingledateselection for our internal representation makes more sense!

Reply:
I don't think for the range one it will make sense to have it be a `MatRangeDateSelection` though. The concept of a minimum range doesn't really make sense

Reply:
hmm I think it clears up the date comparison section and reduces the copied code-- if all the internal dates are stored in the same way it's easier to debug/handle?

Reply:
Ok, for now I'd like to just get this in so we can combine this work with the calendar work, but I do think we might need to revisit this when we create the common base class for `MatDatepickerInput` and `MatDateRangeInput`